File: Chain of Responsibility/Ruby/Caching.rb
require_relative "Handler"
require_relative "Response"

class Caching < Handler
    def initialize()
        @data = {}
    end

    def execute(request)
        if @data[request.route]
            puts "Cached"

            return @data[request.route]
        else
            temp_data = @next_handler.execute(request)
            @data[request.route] = temp_data

            return temp_data
        end
    end
end

File: Chain of Responsibility/Javascript/Caching.js
import Handler from "./Handler.js";

class Caching extends Handler {
  constructor() {
    super();
    this.data = {};
  }

  execute(request) {
    if (this.data[request.route]) {
      console.log("Cached");

      return this.data[request.route];
    } else {
      const temp_data = this.next_handler.execute(request);
      this.data[request.route] = temp_data;

      return temp_data;
    }
  }
}

export default Caching;
